### PR TITLE
[move] Simplify binary version override logic

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

The logic around overriding the binary version output by the compiler was spread out over a number of places, and in general was a bit nasty. This simplifies the logic so the version setting happens in only one place in the code base. 

## Test plan 

Make sure existing tests pass.

### CODE REVIEW

File: external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs
Comment (tnowacki) [inline external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs:168]: Why not with_version? Is this test only code?

File: external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs
Comment (tnowacki) [inline external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:339]: Do we want to just stick this in `ModuleDefinition`? I know it will make this a bigger change, but it feels like the right thing long term...

File: external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs
Comment (tzakian) [reply external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs:168]: This is a convenience method on `CompiledUnit` that uses `serialize_with_version` under the hood, but it's just so you don't need to do the whole `let mut bytes = vec![]; unit.unit.module.serialize_with_version(unit.unit.module.version, &mut bytes); bytes`. So I kinda liked keeping it. But I can remove if you want it gone completely.

File: external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs
Comment (cgswords) [inline external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:442]: `fix` is an overloaded word; maybe set?

File: external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs
Comment (tzakian) [reply external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:339]: Had a look, and sadly this isn't possible :'(

`ModuleDefinition` is in `move-ir-types` so if we added this to that it would create a circular crate dependency + it would really be the wrong place for this function to live I feel.

File: external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs
Comment (tnowacki) [reply external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs:168]: Ah this is on the compiler's unit, not the CompiledModule?

File: external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs
Comment (tnowacki) [inline external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:435]: Hmmmm. Should this be `get_bytecode_version_from_env().or(version)`? 
Or maybe we just nuke `get_bytecode_version_from_env` since there is a compiler flag?

Comment (tzakian) [reply external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:435]: Only reason I kept the env as well is because even though there's a compiler flag the flag is not surfaced up into the package system/the `move build` command. So if you ever wanted to build a package using the existing workflows and specify the version the only way of doing that is with the env variable (at least currently). 

> Should this be get_bytecode_version_from_env().or(version)

I went back and forth on this lol and which should override the other if both are set. I'm happy either way way so fine switching it around if that's preferable.

File: external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs
Comment (tzakian) [reply external-crates/move/crates/move-cli/src/sandbox/commands/publish.rs:168]: yep! This is on `CompiledUnit` :)

File: external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs
Comment (tnowacki) [reply external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:435]: I don't really care. Let's maybe just focus on propagating that flag and killing `get_bytecode_version_from_env`

**Review by tnowacki — APPROVED** [review]
Approved assuming this takes us closer to killing get_bytecode_version_from_env. If we want to just fully kill it in this PR, we can do that too (just let me take a look before landing :))

File: external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs
Comment (tzakian) [reply external-crates/move/crates/move-ir-to-bytecode/src/compiler.rs:435]: I think I'll keep `get_bytecode_version_from_env` at least in this PR since I think there's some nuance that's going to be needed in terms of how we expose the flag out into the build system (i.e., I don't think it should be exposed by default, but maybe only on debug/dev builds etc). Happy to lead the charge on that though (but in a separate PR :) )

**Pull request merged**
